rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -494,20 +489,67 @@ public static StatusExt getStatus(FlowNode node) {
         return nodes;
     }
 
+    protected static String itemFullNameToUrl(String fullName) {
+        StringBuilder output = new StringBuilder();
+        for (String s : fullName.split("/")) {
+            output.append("job/");
+            output.append(s);
+            output.append("/");
+        }
+        return output.toString();
+    }
 
     /** This is used to cover an obscure case where a WorkflowJob is renamed BUT
      *  a previous WorkflowJob existed with cached execution data.
      *  Otherwise the previous job's cached data would be returned.
      **/
     @Extension
     public static class RenameHandler extends ItemListener {
+
+        private void invalidateExecutionCache(WorkflowRun run, CacheExtension ext){
+            try {
+                FlowExecution f = run.getExecution();
+                if (f != null) {
+                    String url = f.getUrl();
+                    ext.getExecutionCache().invalidate(url);
+                }
+            } catch (IOException ioe) {
+                LOGGER.log(Level.WARNING, "Can't get execution URL for WorkflowRun!", ioe);
+            }
+        }
+
         @Override
         public void onLocationChanged(Item item, String oldFullName, String newFullName) {
+            // Better solution: scam through cache, find all instances where oldFullName matches
+            // Replace with newFullName
+            CacheExtension ext = CacheExtension.all().get(0);
+            Cache<String, RunExt> rc = ext.getRunCache();
+
             if (item instanceof WorkflowJob) {
-                // It's a cache, we can rebuild it... we have the technology.
-                CacheExtension ext = CacheExtension.all().get(0);
-                ext.getExecutionCache().invalidateAll();
-                ext.getRunCache().invalidateAll();
+                RunList<WorkflowRun> runs = ((WorkflowJob) item).getBuilds().limit(JobExt.MAX_RUNS_PER_JOB+5);  // Add a few to help invalidate just-completed
```

Comment:
Could you not just look in the cache for entries with the right key, without actually needing to load any `WorkflowRun`s?